File: wavefront.cpp
void vec_print(glm::vec3 vector) {
    printf("%f, %f, %f\n", vector.x, vector.y, vector.z);
}

int parse_indexed_face(const char* line, int* triplet) {
    triplet[0] = 1; triplet[1] = 1; triplet[2] = 1;
    int pos = 0;
    while(line[pos] == 0x20) { pos++; }
    sscanf(&line[pos], "%i", &triplet[0]);
    //forward to just after next slash
    while(line[pos] != 0x2F) { pos++; } pos++;
    if(line[pos] != 0x2F) {
        //if the character isn't a slash then get the value
        sscanf(&line[pos], "%i", &triplet[1]);
        //forward to just after next slash
        while(line[pos] != 0x2F) { pos++; } pos++;
    }
    else {
        //go past the slash
        pos++;
    }
    if(line[pos] != 0x20) {
        sscanf(&line[pos], "%i", &triplet[2]);
    }
    triplet[0]--; triplet[1]--; triplet[2]--; //wavefront obj is 1-indexed
    return pos;
}

glm::vec3 parse_point_string(char* line) {
    //TODO: make this more resilient against instances of multiple characters
    //of whitespace, do this string parsing properly
    int i = 0; while(line[i] == 0x20) { i++; }
    GLfloat x = 0.f; GLfloat y=0.f; GLfloat z=0.f;
    sscanf(&line[i], "%f %f %f", &x, &y, &z);
    return glm::vec3(x, y, z);
}

void compute_face_tb(Face* face, glm::vec3* tangent, glm::vec3* bitangent) {
    glm::vec3 q1; glm::vec3 q2;
    glm::vec3 s1t1; glm::vec3 s2t2;
    Vertex* a = &face->a; Vertex* b = &face->b; Vertex* c = &face->c;
    //q1 = b.v - a.v; q2 = c.v - a.v
    q1 = b->v - a->v; q2 = c->v - a->v;
    //s1t1 = b.uv - a.uv; s2t2 = c.uv - a.uv
    s1t1 = b->uv - a->uv; s2t2 = c->uv - a->uv;
    //if s1t1 == s2t2 #otherwise we get NaN trying to divide infinity
    if(s1t1 == s2t2) {
        //I should probably just do the calculation then re-do it
        //with these inputs if I get NaN
        s1t1 = glm::vec3(1.f, 0.f, 0.f);
        s2t2 = glm::vec3(0.f, 1.f, 0.f);
    }

    GLfloat divisor = (s1t1.x*s2t2.y) - (s1t1.y*s2t2.x);

    tangent->x  = ((s2t2.y * q1.x) - (s1t1.y * q2.x)) / divisor;
    tangent->y  = ((s2t2.y * q1.y) - (s1t1.y * q2.y)) / divisor;
    tangent->z  = ((s2t2.y * q1.z) - (s1t1.y * q2.z)) / divisor;
    normalize(tangent);

    bitangent->x = ((s1t1.x * q2.x) - (s2t2.x*q1.x)) / divisor;
    bitangent->y = ((s1t1.x * q2.y) - (s2t2.x*q1.y)) / divisor;
    bitangent->z = ((s1t1.x * q2.z) - (s2t2.x*q1.z)) / divisor;
    normalize(bitangent);

    return;
}

int load_face_array(FILE* file, const char* label, Indexed_Face** faces) {
    char buffer[255]; int item_count = 0;
    while(fgets(buffer, sizeof(buffer), file) != NULL) {
        if(strncmp(buffer, label, strlen(label)) == 0) { item_count++; }
    }

    *faces = (Indexed_Face*)malloc(sizeof(Indexed_Face)*item_count);

    int i = 0; fseek(file, 0, SEEK_SET);
    int pos; int* triplet = (int*)malloc(sizeof(int)*3);
    Indexed_Face face;
    while(fgets(buffer, sizeof(buffer), file) != NULL) {
        if(strncmp(buffer, label, strlen(label)) == 0) {
            pos = strlen(label);
            pos += parse_indexed_face(&buffer[pos], triplet);
            face.a_v_index = triplet[0];
            face.a_uv_index = triplet[1];
            face.a_n_index = triplet[2];
            while(buffer[pos] != 0x20) { pos++; }
            pos += parse_indexed_face(&buffer[pos], triplet);
            face.b_v_index = triplet[0];
            face.b_uv_index = triplet[1];
            face.b_n_index = triplet[2];
            while(buffer[pos] != 0x20) { pos++; }
            pos += parse_indexed_face(&buffer[pos], triplet);
            face.c_v_index = triplet[0];
            face.c_uv_index = triplet[1];
            face.c_n_index = triplet[2];
            (*faces)[i] = face;
            i++;
        }
    }

    fseek(file, 0, SEEK_SET);
    return item_count;
}

int load_vec3_array(FILE* file, const char* label, glm::vec3** items) {
    char buffer[255]; int item_count = 0;
    while(fgets(buffer, sizeof(buffer), file) != NULL) {
        if(strncmp(buffer, label, strlen(label)) == 0) { item_count++; }
    }

    if(item_count == 0) {
        item_count = 1;
        *items = (glm::vec3*)malloc(sizeof(glm::vec3)*item_count);
        (*items)[0] = glm::vec3(0.f, 0.f, 0.f);
        fseek(file, 0, SEEK_SET);
        return 0;
    }

    *items = (glm::vec3*)malloc(sizeof(glm::vec3)*item_count);

    int i = 0; fseek(file, 0, SEEK_SET);
    while(fgets(buffer, sizeof(buffer), file) != NULL) {
        if(strncmp(buffer, label, strlen(label)) == 0) {
            (*items)[i] = parse_point_string(buffer+strlen(label));
            i++;
        }
    }

    fseek(file, 0, SEEK_SET);
    return item_count;
}

bool load_model(const char* model_name, Model* model) {
    //Load file
    model->asset_path = construct_asset_path("models", model_name, "obj");
    char buffer[255]; FILE* file = fopen(model->asset_path, "r");

    if(file == NULL) {
        message_log("Error loading file-", model->asset_path); return false;
    }

    glm::vec3* vertices;
    int vertex_count = load_vec3_array(file, "v ", &vertices);
    glm::vec3* uvs;
    int uv_count = load_vec3_array(file, "vt", &uvs);
    glm::vec3* normals;
    int normal_count = load_vec3_array(file, "vn", &normals);
    Indexed_Face* indexed_faces;
    int face_count = load_face_array(file, "f ", &indexed_faces);

    fclose(file);

    //Assemble Faces
    Face* faces = (Face*)malloc(sizeof(Face)*face_count);
    Indexed_Face indexed_face;
    Face face;
    for(int i = 0; i < face_count; i++) {
        indexed_face = indexed_faces[i];
        face.a.v = vertices[indexed_face.a_v_index];
        face.a.uv = uvs[indexed_face.a_uv_index];
        face.a.n = normals[indexed_face.a_n_index];
        face.a.t = glm::vec3(0.f, 0.f, 0.f);
        face.a.b = glm::vec3(0.f, 0.f, 0.f);
        face.b.v = vertices[indexed_face.b_v_index];
        face.b.uv = uvs[indexed_face.b_uv_index];
        face.b.n = normals[indexed_face.b_n_index];
        face.b.t = glm::vec3(0.f, 0.f, 0.f);
        face.b.b = glm::vec3(0.f, 0.f, 0.f);
        face.c.v = vertices[indexed_face.c_v_index];
        face.c.uv = uvs[indexed_face.c_uv_index];
        face.c.n = normals[indexed_face.c_n_index];
        face.c.t = glm::vec3(0.f, 0.f, 0.f);
        face.c.b = glm::vec3(0.f, 0.f, 0.f);
        faces[i] = face;
    }
    model->faces = faces;
    model->face_count = face_count;

    //bail if there's no UV's and thus we don't need tangents/bitangents
    if(uv_count == 0) {
        free(vertices); free(uvs); free(normals); free(indexed_faces);
        return true;
    }

    //initialize tangents/bitangents
    float* tangent_uses = (float*)malloc(sizeof(float)*vertex_count);
    glm::vec3* tangents = (glm::vec3*)malloc(sizeof(glm::vec3)*vertex_count);
    glm::vec3* bitangents = (glm::vec3*)malloc(sizeof(glm::vec3)*vertex_count);

    for(int i = 0; i < vertex_count; i++) {
        tangents[i] = glm::vec3(0.f, 0.f, 0.f);
        bitangents[i] = glm::vec3(0.f, 0.f, 0.f);
        tangent_uses[i] = 0.f;
    }

    glm::vec3 tangent; glm::vec3 bitangent;
    for(int i = 0; i < face_count; i++) {
        //Compute tangent/bitangent, prep for averaging
        compute_face_tb(&faces[i], &tangent, &bitangent);

        tangents[indexed_faces[i].a_v_index] += tangent;
        bitangents[indexed_faces[i].a_v_index] += bitangent;
        tangents[indexed_faces[i].b_v_index] += tangent;
        bitangents[indexed_faces[i].b_v_index] += bitangent;
        tangents[indexed_faces[i].c_v_index] += tangent;
        bitangents[indexed_faces[i].c_v_index] += bitangent;

        tangent_uses[indexed_faces[i].a_v_index] += 1.f;
        tangent_uses[indexed_faces[i].b_v_index] += 1.f;
        tangent_uses[indexed_faces[i].c_v_index] += 1.f;
    }


    //compute average tangent/bitangent for each vertex
    float uses;
    for(int i = 0; i < vertex_count; i++) {
        uses = tangent_uses[i];
        tangents[i] = tangents[i] / uses;
        bitangents[i] = bitangents[i] / uses;

        normalize(&tangents[i]);
        normalize(&bitangents[i]);
    }

    //attach tangents/bitangents to faces
    int vertex_index;
    for(int i = 0; i < face_count; i++) {
        faces[i].a.t = tangents[indexed_faces[i].a_v_index];
        faces[i].a.b = bitangents[indexed_faces[i].a_v_index];
        faces[i].b.t = tangents[indexed_faces[i].b_v_index];
        faces[i].b.b = bitangents[indexed_faces[i].b_v_index];
        faces[i].c.t = tangents[indexed_faces[i].c_v_index];
        faces[i].c.b = bitangents[indexed_faces[i].c_v_index];
    }

    free(vertices); free(uvs); free(normals); free(tangents); free(bitangents);
    free(indexed_faces); free(tangent_uses);
    return true;
}

File: wavefront.rb
require 'matrix'

class Point
    def self.from_array(array)
        return Point.new(array[0], array[1], array[2]);
        return self;
    end

    def u
        return self.x;
    end

    def v
        return self.y;
    end

    def +(other)
        return Point.new(self.x + other.x, self.y + other.y, self.z + other.z)
    end

    def -(other)
        return Point.new(self.x - other.x, self.y - other.y, self.z - other.z)
    end
end

def compute_face_tb(face) #tangent/bitangent
    a, b, c = face
    q1 = b.v - a.v; q2 = c.v - a.v
    s1t1 = b.uv - a.uv; s2t2 = c.uv - a.uv
    if s1t1 == s2t2 #otherwise we get NaN trying to divide infinity
        s1t1 = Point.new(1.0, 0.0, 0.0)
        s2t2 = Point.new(0.0, 1.0, 0.0)
    end
    st_matrix = Matrix[ [s2t2.v, -1*s1t1.v],
                        [-1*s2t2.u, s1t1.u] ]
    q_matrix = Matrix[ [q1.x, q1.y, q1.z],
                       [q2.x, q2.y, q2.z] ]
    tb_matrix = (st_matrix * q_matrix)/(s1t1.u*s2t2.v - s1t1.v*s2t2.u)
    t = Point.from_array(tb_matrix.row(0)).normalize
    b = Point.from_array(tb_matrix.row(1)).normalize
    return [t,b]
end

class Indexed_Vertex
    #Shim class to load indexed vertex information in before processing and
    #sending to C_Vertex which requires proper Point* objects instead of ints
    attr_accessor :v; attr_accessor :uv; attr_accessor :normal;
    def initialize(geometric_vertex, texture_coordinate, normal)
        @v = geometric_vertex; @uv = texture_coordinate; @normal = normal;
    end
end

class Wavefront
#TODO: no reason for this to be a class anymore, should create a Face struct in
#C and just return a Ruby array of C_Faces that gets converted on load
    attr_reader :faces

    def initialize(faces, vertices, uvs, normals)
        @vertices = vertices
        @uvs = uvs
        @normals = normals
        @tangents = Array.new(vertices.length){[]}
        @bitangents = Array.new(vertices.length){[]}

        #compute tangents/bitangents for tangent space normal mapping
        for indexed_face in faces
            vertex_indices = indexed_face.map{ |vertex| vertex.v }
            face = indexed_face.map{ |vertex|
                    vertex = vertex.dup
                    vertex.v = @vertices[vertex.v]
                    vertex.uv = @uvs[vertex.uv]
                    vertex.normal = @normals[vertex.normal]
                    vertex
                    }
            tangent, bitangent = compute_face_tb(face)
            for index in vertex_indices
                @tangents[index] << tangent
                @bitangents[index] << bitangent
            end
        end
        @faces = faces

        #average face tangent/bitangets to get t/b at individual vertices
        @tangents.map!{|group| group.inject(&:+).scale_by_factor(1.0/group.length) }
        @bitangents.map!{|group| group.inject(&:+).scale_by_factor(1.0/group.length) }

        #Send everything over to C
        @faces.map!{ |indexed_face| self.build_face(indexed_face) }
        return @faces
    end

    def build_face(indexed_face)
        return C_Face.new( self.build_vertex(indexed_face[0]),
            self.build_vertex(indexed_face[1]), self.build_vertex(indexed_face[2]))
    end

    def build_vertex(vertex)
        v = @vertices[vertex.v]
        uv = @uvs[vertex.uv]
        normal = @normals[vertex.normal]
        tangent = @tangents[vertex.v]
        bitangent = @bitangents[vertex.v]
        return C_Vertex.new(v, uv, normal, tangent, bitangent)
    end

	def self.from_file(file_path)
        #log("Loading model: #{file_path}")
        faces, vertices, uvs, normals  = [], [], [], []

        lines = open(file_path).readlines.map!(&:strip)
        for line in lines
            parts = line.split(" ")
            if parts[0] == "v"
                x, y, z = parts[1..-1].map(&:to_f)
                vertices << Point.new(x, y, z)

            elsif parts[0] == "vt"
                x, y, z = parts[1..-1].map(&:to_f)
                uvs << Point.new(x, y, z)

            elsif parts[0] == "vn"
                x, y, z = parts[1..-1].map(&:to_f)
                normals << Point.new(x, y, z)

            elsif parts[0] == "f"
                v, vt, vn = parts[1].split("/").map{ |index| index.to_i - 1}
                a = Indexed_Vertex.new(v, vt, vn)
                v, vt, vn = parts[2].split("/").map{ |index| index.to_i - 1}
                b = Indexed_Vertex.new(v, vt, vn)
                v, vt, vn = parts[3].split("/").map{ |index| index.to_i - 1}
                c = Indexed_Vertex.new(v, vt, vn)
                faces << [a, b, c]
            end

        end
        return Wavefront.new(faces, vertices, uvs, normals)
    end
end

def face_to_screen(face, view_matrix, screen_center)
    face = face.dup
    face[0].v = face[0].v.dup.apply_matrix!(view_matrix).to_screen!(screen_center)
    face[1].v = face[1].v.dup.apply_matrix!(view_matrix).to_screen!(screen_center)
    face[2].v = face[2].v.dup.apply_matrix!(view_matrix).to_screen!(screen_center)
    face = face.sort_by(&:v)
    return face
end

def compute_face_normal(face)
    a, b, c = face
    return ((b.v - a.v).cross_product(c.v - a.v)).normalize!
end

